File: Services/dotnet-services/Core/OmniCure.Core.Application/Interfaces/Caching/IRedisSearchService.cs
namespace OmniCure.Core.Application.Interfaces.Caching;

public record SearchDocument(string Id, Dictionary<string, string> Fields);

public record SearchResponse(long TotalResults, List<SearchDocument> Documents);

public interface IRedisSearchService
{
    Task CreateIndexAsync(string indexName, string prefix, Action<ISchemaBuilder> schemaBuilder);
    Task DropIndexAsync(string indexName, bool deleteDocuments = false);
    Task<bool> IndexExistsAsync(string indexName);
    Task ReindexAsync(string indexName, string prefix, Action<ISchemaBuilder> schemaBuilder);

    Task<SearchResponse> SearchAsync(string indexName, string query, int offset = 0, int limit = 10);
    Task<SearchResponse> SearchAsync(string indexName, string query, string? sortBy = null, bool ascending = true, int offset = 0, int limit = 10, params string[] returnFields);

    Task SetHashAsync(string key, Dictionary<string, object> fields, TimeSpan? expiry = null);
    Task DeleteHashAsync(string key);
}

public interface ISchemaBuilder
{
    ISchemaBuilder AddTextField(string name, double weight = 1.0, bool sortable = false, bool noStem = false);
    ISchemaBuilder AddTagField(string name, string separator = ",", bool sortable = false);
    ISchemaBuilder AddNumericField(string name, bool sortable = false);
    ISchemaBuilder AddGeoField(string name);
    ISchemaBuilder AddVectorField(string name, string algorithm, int dim, string distanceMetric);
}

File: Services/java-services/Core/CoreApplication/src/main/java/ai/omnicure/core/application/cache/IRedisSearchService.java
package ai.omnicure.core.application.cache;

import java.util.List;
import java.util.Map;

public interface IRedisSearchService {

    void createIndex(String indexName, String prefix, java.util.function.Consumer<SchemaBuilder> schemaBuilder);

    void dropIndex(String indexName, boolean deleteDocuments);

    boolean indexExists(String indexName);

    void reindex(String indexName, String prefix, java.util.function.Consumer<SchemaBuilder> schemaBuilder);

    SearchResponse search(String indexName, String query, int offset, int limit);

    SearchResponse search(String indexName, String query, String sortBy, boolean ascending, int offset, int limit, String... returnFields);

    void setHash(String key, Map<String, Object> fields, java.time.Duration expiry);

    void deleteHash(String key);

    record SearchDocument(String id, Map<String, String> fields) {}

    record SearchResponse(long totalResults, List<SearchDocument> documents) {}

    interface SchemaBuilder {
        SchemaBuilder addTextField(String name, double weight, boolean sortable, boolean noStem);
        SchemaBuilder addTagField(String name, String separator, boolean sortable);
        SchemaBuilder addNumericField(String name, boolean sortable);
        SchemaBuilder addGeoField(String name);
        SchemaBuilder addVectorField(String name, String algorithm, int dim, String distanceMetric);
    }
}
